File: packages/tracecast-ts/real_examples/langchainAgent.ts
import path from "path";
import dotenv from "dotenv";
dotenv.config({ path: path.resolve(__dirname, "../../../.env") });

import { ChatOpenAI } from "@langchain/openai";
import { ChatPromptTemplate } from "@langchain/core/prompts";
import { Tracer, JsonFileExporter, DictExporter } from "../src";
import { TraceCastCallback } from "../src/integrations/langchain";

const tracer = new Tracer({
  exporters: [
    new JsonFileExporter("./traces/ts_langchain_traces.jsonl"),
    new DictExporter({
      onTrace: (d: any) =>
        console.log(`[ANALYTICS] ${d.name} cost=$${d.costUsd.toFixed(4)}`),
    }),
  ],
  logging: true,
  logPrefix: "ts_langchain_agent",
});
const cb = new TraceCastCallback(tracer);

async function translate(text: string, targetLang: string): Promise<string> {
  const llm = new ChatOpenAI({ model: "gpt-4o-mini", temperature: 0 });
  const prompt = ChatPromptTemplate.fromMessages([
    ["system", `Você é um tradutor especializado. Traduza o texto para ${targetLang}. Responda apenas com a tradução.`],
    ["user", "{text}"],
  ]);
  const chain = prompt.pipe(llm as any) as any;

  return tracer.trace(
    "langchain_translate",
    async () => {
      const result = await chain.invoke({ text }, { callbacks: [cb] });
      return result.content as string;
    },
    { userId: "user_pedro", metadata: { targetLang, framework: "langchain" } },
  );
}

if (require.main === module) {
  const examples = [
    { text: "Hello, how are you?", lang: "Português" },
    { text: "The quick brown fox jumps over the lazy dog", lang: "Espanhol" },
  ];

  (async () => {
    for (const { text, lang } of examples) {
      const result = await translate(text, lang);
      console.log(`\n[${lang}] ${text}\n→ ${result}`);
    }
  })().catch(console.error);
}

File: packages/tracecast-py/real_examples/langchain_agent.py
import logging
from pathlib import Path

try:
    from dotenv import load_dotenv
    load_dotenv(Path(__file__).parents[3] / ".env")
except ImportError:
    pass

from langchain.agents import AgentExecutor, create_tool_calling_agent
from langchain.tools import tool
from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
from langchain_openai import ChatOpenAI
from tracecast import Tracer
from tracecast.exporters import DictExporter, JsonFileExporter
from tracecast.integrations.langchain import TraceCastCallback

logging.basicConfig(level=logging.INFO, format="%(message)s")

tracer = Tracer(
    exporters=[
        JsonFileExporter("./traces/langchain_traces.jsonl"),
        DictExporter(on_trace=lambda d: print(f"[ANALYTICS] {d['name']} cost=${d['cost_usd']:.4f}")),
    ],
    logging=True,
    log_prefix="financial_agent",
)
callback = TraceCastCallback(tracer=tracer)


@tool
def get_stock_price(ticker: str) -> str:
    prices = {"PETR4": 38.75, "VALE3": 65.20, "ITUB4": 32.10}
    price = prices.get(ticker.upper(), 0.0)
    return f"{ticker.upper()}: R$ {price:.2f}"


@tool
def calculate_investment_return(principal: float, rate: float, months: int) -> str:
    total = principal * ((1 + rate / 100) ** months)
    return f"Principal: R$ {principal:.2f} → Total: R$ {total:.2f} em {months} meses a {rate}% a.m."

llm = ChatOpenAI(model="gpt-4o-mini", temperature=0, stream_usage=True)
tools = [get_stock_price, calculate_investment_return]
prompt = ChatPromptTemplate.from_messages([
    ("system", "Você é um assistente financeiro brasileiro. Responda sempre em português."),
    MessagesPlaceholder("chat_history", optional=True),
    ("user", "{input}"),
    MessagesPlaceholder("agent_scratchpad"),
])
agent = create_tool_calling_agent(llm, tools, prompt)
executor = AgentExecutor(agent=agent, tools=tools, verbose=False)


def analyze_portfolio(question: str) -> str:
    with tracer.trace("financial_agent_run", user_id="user_pedro"):
        result = executor.invoke({"input": question}, config={"callbacks": [callback]})
        return result["output"]


if __name__ == "__main__":
    questions = [
        "Qual a cotação de PETR4?",
        "Rendimento de R$ 10.000 a 0.8% por 12 meses",
        "Compare PETR4 e VALE3 e calcule rendimento de R$ 5.000 a 0.9% por 6 meses",
    ]
    for q in questions:
        answer = analyze_portfolio(q)
        print(f"\nPergunta: {q}\nResposta: {answer}\n")
